File: exp12/FeatTracking.cpp
#include <bits/stdc++.h>
#include <opencv2/opencv.hpp>
#include <opencv2/tracking.hpp>

using namespace std;
using namespace cv;

const char *videoPath = "/home/liyang/视频/AvA_bunny.mp4";

Mat img, img_gray, img_pre, img_pre_gray, screenshot, screenshot_grey, target, merged;
Rect bbox, bbox_copy;//NOLINT 检测框
bool captured = false;

vector<Point2f> point1, point2, point_copy;
vector<uchar> status;
vector<float> err;

void showImage(Mat &image, const char *windowName, int location_x = 0, int location_y = 0) {
    //当窗口关闭时，getWindowProperty将返回-1
    if (getWindowProperty(windowName, WND_PROP_AUTOSIZE) == -1) {
        namedWindow(windowName, WINDOW_AUTOSIZE);
        moveWindow(windowName, location_x, location_y);
    }
    imshow(windowName, image);
}

void targetTracking() {
    Ptr<Tracker> tracker = TrackerKCF::create();
    VideoCapture video(videoPath);
    if (!video.isOpened()) {
        cout << "video open error." << '\n';
        return;
    }

    double fps = video.get(CAP_PROP_FPS); //获取视频帧率
    double pauseTime = 1000 / fps; //两幅画面中间间隔
    while (video.read(img)) {
        auto t = (double) getTickCount();
        if (waitKey(pauseTime) == 27) {//NOLINT Esc键按下暂停，选取目标
            bbox = selectROI("video", img, false);
            bbox_copy = bbox;
            cout << bbox.x << " " << bbox.y << " " << bbox.width << " " << bbox.height << '\n';
            target = img(bbox);

            img.copyTo(screenshot);
            rectangle(screenshot, bbox, Scalar(0, 0, 255), 2);
            tracker->init(screenshot, bbox);//跟踪器初始化

            img.copyTo(img_pre);
            cvtColor(img_pre, img_pre_gray, COLOR_BGR2GRAY);
            goodFeaturesToTrack(img_pre_gray(bbox), point1, 100, 0.01, 10, Mat());
            point_copy = point1;
            for (auto &p:point1) circle(screenshot(bbox), p, 1, Scalar(255, 0, 0), 2);

            imwrite("/home/liyang/视频/screenshot.png", screenshot);
            imwrite("/home/liyang/视频/target.png", target);
            captured = true;
        }
        if (!captured) {
            showImage(img, "video");
            continue;
        }
        if (tracker->update(img, bbox)) {
            rectangle(img, bbox, (0, 0, 255), 2);

            cvtColor(img, img_gray, COLOR_BGR2GRAY);
            calcOpticalFlowPyrLK(img_pre_gray, img_gray, point1, point2, status, err, Size(50, 50), 3);
            for (auto &p:point2) circle(img(bbox), p, 1, Scalar(255, 0, 0), 2);

            //合并截图和当前帧
            int rows = img.rows, cols = img.cols;
            merged.create(rows, cols * 2, img.type());
            img.copyTo(merged(Rect(0, 0, cols, rows)));
            screenshot.copyTo(merged(Rect(cols, 0, cols, rows)));
            auto gap = getTickFrequency() / ((double) getTickCount() - t);
            putText(merged, "FPS : " + to_string(int(gap)), Point(100, 50), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(50, 170, 50),
                    2);
            for (int i = 0; i < point2.size(); i++) {
                Point2f pt1(point2[i].x + bbox.x, point2[i].y + bbox.y), pt2(point_copy[i].x + cols + bbox_copy.x,
                                                                             point_copy[i].y + bbox_copy.y);
                line(merged, pt1, pt2, Scalar(0, 255, 0));
            }
            showImage(merged, "video");

            swap(point1, point2);
            img_pre_gray = img_gray.clone();
        } else {
            int rows = img.rows, cols = img.cols;
            merged.create(rows, cols * 2, img.type());
            img.copyTo(merged(Rect(0, 0, cols, rows)));
            screenshot.copyTo(merged(Rect(cols, 0, cols, rows)));
            auto gap = getTickFrequency() / ((double) getTickCount() - t);
            putText(merged, "FPS : " + to_string(int(gap)), Point(100, 50), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(50, 170, 50),
                    2);
            showImage(merged, "video");
        }
    }
    video.release();
}

int main() {
    ios::sync_with_stdio(false);

    namedWindow("video");
    moveWindow("video", 400, 200);

    targetTracking();

    waitKey(0);
    destroyAllWindows();

    return 0;
}

File: exp12/FeatTracking.py
import numpy as np
import cv2 as cv
import os
import argparse

global screenshot, roi, target, bbox
global k_roi, d_roi
captured = False
blue, green, red = (255, 0, 0), (0, 255, 0), (0, 0, 255)


def show_image(img, win_name: str, location_x=0, location_y=0) -> None:
    if cv.getWindowProperty(win_name, cv.WND_PROP_AUTOSIZE) == -1:
        cv.namedWindow(win_name)
        cv.moveWindow(win_name, location_x, location_y)
    cv.imshow(win_name, img)


def merge_image(img1, img2):
    assert img1.shape == img2.shape
    row, col, c = img1.shape
    merge = np.zeros((row, col * 2, c), dtype='uint8')
    merge[:row, :col] = img1
    merge[:row, col:] = img2
    return merge


def feature_detect(img):
    def keypoint2point(keypoint):
        point = np.zeros(len(keypoint) * 2, np.float32)
        for i in range(len(keypoint)):
            point[i * 2] = keypoint[i].pt[0]
            point[i * 2 + 1] = keypoint[i].pt[1]
        point = point.reshape(-1, 2)
        return point

    img_gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY) if img.shape[-1] == 3 else img
    surf = cv.xfeatures2d.SURF_create(2000)
    key_pt, desc = surf.detectAndCompute(img_gray, None)
    return key_pt, keypoint2point(key_pt), desc


def get_match_image(img_left, kps_left, desc_left, img_right, kps_right, desc_right):
    matcher = cv.DescriptorMatcher_create("BruteForce")
    raw_matches = matcher.knnMatch(desc_left, desc_right, 2)
    good = []
    for m, n in raw_matches:
        if m.distance < 0.7 * n.distance:
            good.append(m)
    return cv.drawMatches(img_left, kps_left, img_right, kps_right, good, None)


def target_tracking(file: str) -> None:
    def get_roi(img, box):
        ret = np.zeros(img.shape, dtype='uint8')
        p1, p2 = bbox2point(box)
        ret[p1[1]:p2[1], p1[0]:p2[0]] = img[p1[1]:p2[1], p1[0]:p2[0]]
        return ret

    def bbox2point(box):
        p1 = (int(box[0]), int(box[1]))
        p2 = (int(box[0] + box[2]), int(box[1] + box[3]))
        return p1, p2

    global screenshot, roi, target, bbox, captured
    global k_roi, d_roi
    tracker = cv.TrackerKCF_create()
    video = cv.VideoCapture(file)
    pause_time = int(1000 / video.get(cv.CAP_PROP_FPS))
    while True:
        ok, frame = video.read()
        if not ok:
            break
        t = cv.getTickCount()
        if cv.waitKey(pause_time) == 27:  # press Esc to select target
            bbox = cv.selectROI('video', frame, False)
            pt1, pt2 = bbox2point(bbox)

            screenshot = frame
            cv.rectangle(screenshot, pt1, pt2, red, 2)
            roi = get_roi(frame, bbox)
            k_roi, _, d_roi = feature_detect(roi)
            target = frame[pt1[1]:pt2[1], pt1[0]:pt2[0]]
            show_image(target, 'target')

            ok = tracker.init(frame, bbox)
            captured = True
        if not captured:
            show_image(frame, 'video')
            continue

        ok, bbox = tracker.update(frame)
        if ok:
            cv.rectangle(frame, bbox2point(bbox)[0], bbox2point(bbox)[1], blue, 2)
            roi_frame = get_roi(frame, bbox)
            k_frame, _, d_frame = feature_detect(roi_frame)
            merge = get_match_image(frame, k_frame, d_frame, screenshot, k_roi, d_roi)
        else:
            merge = merge_image(frame, screenshot)
        fps = cv.getTickFrequency() / (cv.getTickCount() - t)
        cv.putText(merge, f'FPS: {int(fps)}', (100, 50), cv.FONT_HERSHEY_SIMPLEX, 0.75, (50, 170, 50), 2)

        show_image(merge, 'video')


def main():
    video_path = os.path.join(args.video_dir, f'{args.video_name}.mp4')
    assert os.path.exists(video_path), 'file not exists.'

    cv.namedWindow('video')
    cv.moveWindow('video', 400, 200)
    target_tracking(video_path)

    cv.waitKey(0)
    cv.destroyAllWindows()


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--video_dir', type=str, default='/home/liyang/视频')
    parser.add_argument('--video_name', type=str, default='AvA_bunny')
    args = parser.parse_args()
    print(args)

    main()
